combine filter: fine list handling (option b)

tchernomax opened the pull request:
##### SUMMARY

see:

* #57623
* https://github.com/ansible/community/issues/474#issuecomment-500994513
* https://github.com/ansible/community/issues/474#issuecomment-501765541
* fixes #59386
* fixes #17714
Solution (b)

args:

* recursive: same as before
* list_merge:
  * keep: same as before
  * append, prepend: list are appended
  * append_np, prepend_np: only elements that isn't already present in left list are append (_np stands for "non present")

README: will come if this implementation is validated.

I didn't implement 'uniq' as I think 'append_np' and 'prepend_np' are more useful, and are equivalent to unique if list are already filtered from duplication.

non backward compatible: `[dict1, [dict2]] | combine` doesn't work anymore, because I thought it shouldn't. If you disagree, `flatten(terms, levels=1) → flatten(terms)` and it's fixed.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

component =lib/ansible/plugins/filter/core.py

##### ADDITIONAL INFORMATION

```
max@mde-oxalide % cat toto.yml
- hosts: localhost
  gather_facts: false
  vars:
    toto_default:
      a:
        a:
          a: default_value
          b: default_value
      b:
        - 1
        - 1
        - 2
    toto:
      a:
        a:
          b: toto_value
          c: toto_value
      b:
        - 2
        - 3
        - 3
        - 4: toto_value
  tasks:
    - name: recursive=False, list_merge='keep'
      debug:
        msg: "{{toto_default|toto_combine(toto, recursive=False, list_merge='keep')}}"
    - name: recursive=False, list_merge='append'
      debug:
        msg: "{{toto_default|toto_combine(toto, recursive=False, list_merge='append')}}"
    - name: recursive=False, list_merge='prepend'
      debug:
        msg: "{{toto_default|toto_combine(toto, recursive=False, list_merge='prepend')}}"
    - name: recursive=False, list_merge='append_np'
      debug:
        msg: "{{toto_default|toto_combine(toto, recursive=False, list_merge='append_np')}}"
    - name: recursive=False, list_merge='prepend_np'
      debug:
        msg: "{{toto_default|toto_combine(toto, recursive=False, list_merge='prepend_np')}}"

    - name: recursive=True, list_merge='keep'
      debug:
        msg: "{{toto_default|toto_combine(toto, recursive=True, list_merge='keep')}}"
    - name: recursive=True, list_merge='append'
      debug:
        msg: "{{toto_default|toto_combine(toto, recursive=True, list_merge='append')}}"
    - name: recursive=True, list_merge='prepend'
      debug:
        msg: "{{toto_default|toto_combine(toto, recursive=True, list_merge='prepend')}}"
    - name: recursive=True, list_merge='append_np'
      debug:
        msg: "{{toto_default|toto_combine(toto, recursive=True, list_merge='append_np')}}"
    - name: recursive=True, list_merge='prepend_np'
      debug:
        msg: "{{toto_default|toto_combine(toto, recursive=True, list_merge='prepend_np')}}"
```

```
PLAY [localhost] ***********************************************************************

TASK [recursive=False, list_merge='keep'] **********************************************
ok: [localhost] => {
    "msg": {
        "a": {
            "a": {
                "b": "toto_value",
                "c": "toto_value"
            }
        },
        "b": [
            2,
            3,
            3,
            {
                "4": "toto_value"
            }
        ]
    }
}

TASK [recursive=False, list_merge='append'] ********************************************
ok: [localhost] => {
    "msg": {
        "a": {
            "a": {
                "b": "toto_value",
                "c": "toto_value"
            }
        },
        "b": [
            1,
            1,
            2,
            2,
            3,
            3,
            {
                "4": "toto_value"
            }
        ]
    }
}

TASK [recursive=False, list_merge='prepend'] *******************************************
ok: [localhost] => {
    "msg": {
        "a": {
            "a": {
                "b": "toto_value",
                "c": "toto_value"
            }
        },
        "b": [
            1,
            1,
            2,
            2,
            3,
            3,
            {
                "4": "toto_value"
            }
        ]
    }
}

TASK [recursive=False, list_merge='append_np'] *****************************************
ok: [localhost] => {
    "msg": {
        "a": {
            "a": {
                "b": "toto_value",
                "c": "toto_value"
            }
        },
        "b": [
            2,
            3,
            3,
            {
                "4": "toto_value"
            },
            1,
            1
        ]
    }
}

TASK [recursive=False, list_merge='prepend_np'] ****************************************
ok: [localhost] => {
    "msg": {
        "a": {
            "a": {
                "b": "toto_value",
                "c": "toto_value"
            }
        },
        "b": [
            1,
            1,
            2,
            3,
            3,
            {
                "4": "toto_value"
            }
        ]
    }
}

TASK [recursive=True, list_merge='keep'] ***********************************************
ok: [localhost] => {
    "msg": {
        "a": {
            "a": {
                "a": "default_value",
                "b": "toto_value",
                "c": "toto_value"
            }
        },
        "b": [
            2,
            3,
            3,
            {
                "4": "toto_value"
            }
        ]
    }
}

TASK [recursive=True, list_merge='append'] *********************************************
ok: [localhost] => {
    "msg": {
        "a": {
            "a": {
                "a": "default_value",
                "b": "toto_value",
                "c": "toto_value"
            }
        },
        "b": [
            1,
            1,
            2,
            2,
            3,
            3,
            {
                "4": "toto_value"
            }
        ]
    }
}

TASK [recursive=True, list_merge='prepend'] ********************************************
ok: [localhost] => {
    "msg": {
        "a": {
            "a": {
                "a": "default_value",
                "b": "toto_value",
                "c": "toto_value"
            }
        },
        "b": [
            1,
            1,
            2,
            2,
            3,
            3,
            {
                "4": "toto_value"
            }
        ]
    }
}

TASK [recursive=True, list_merge='append_np'] ******************************************
ok: [localhost] => {
    "msg": {
        "a": {
            "a": {
                "a": "default_value",
                "b": "toto_value",
                "c": "toto_value"
            }
        },
        "b": [
            2,
            3,
            3,
            {
                "4": "toto_value"
            },
            1,
            1
        ]
    }
}

TASK [recursive=True, list_merge='prepend_np'] *****************************************
ok: [localhost] => {
    "msg": {
        "a": {
            "a": {
                "a": "default_value",
                "b": "toto_value",
                "c": "toto_value"
            }
        },
        "b": [
            1,
            1,
            2,
            3,
            3,
            {
                "4": "toto_value"
            }
        ]
    }
}

PLAY RECAP *****************************************************************************
localhost                  : ok=10   changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

Akasurde commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -294,25 +293,93 @@ def mandatory(a):
-    else:
-        return dict(itertools.chain(*map(iteritems, dicts)))
+    recursive = kwargs.pop('recursive', False)
+    list_merge = kwargs.pop('list_merge', 'keep')
```
Is `list_merge` new argument ? If yes, then it would be nice to add documentation, examples, changelog entry for this change.

Akasurde replied:
Also, test case for this change.

tchernomax replied:
@Akasurde yes `list_merge` is a new argument.

This feature (fine handling of list merging during combine) is still in discussed. we will talk about it next Tuesday. That's why I didn't bother to write documentation nor tests.
I will do it after the implementation get validated.

> README: will come if this implementation is validated.

Akasurde replied:
OK, I didn't know this.

bcoca commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -294,25 +293,93 @@ def mandatory(a):
+    recursive_check_defined(dicts)
+
+    # function that will be use to merge 2 dicts
+    def _combine_two_dicts(x, y):
```
you seem to be re-implementing merge_hash, just update the function instead

tchernomax replied:
done

tchernomax commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -294,25 +294,30 @@ def mandatory(a):
+    # merge all the dicts so that the dict at the end of the array have precedence
+    # over the dict at the beginning
+    result = dictionaries[0]
+    for dictionary in dictionaries[1:]:
```
@bcoca I think it would be better to build the result from right to left as there is a huge probability that the leftmost dict will be the bigger one (as the leftmost will be the "default" and the other one will be the "patches"), and merge_hash create a copy of it's first argument.
But it add complexity. So I prefer to ask you before I made the change: should I build the result from left to right ? or right to left ?

tchernomax commented:
@ping bcoca (I made the change you asked for)

jamescassell commented on lib/ansible/utils/vars.py:
```diff
@@ -92,32 +92,90 @@ def combine_vars(a, b):
-            result[k] = v
-
-    return result
+    if list_merge not in ('keep', 'append', 'prepend', 'append_np', 'prepend_np'):
```
might want a `replace` similar to `keep`, and bikeshedding on the names

tchernomax replied:
@jamescassell done

Commit:
combine filter: fine list handling

Commit:
merge_hash: unit tests

Commit:
combine filter: integration tests

Commit:
combine filter: documentation

tchernomax commented:
ready_for_review

felixfontein commented on docs/docsite/rst/user_guide/playbooks_filters.rst:
```diff
@@ -890,36 +890,217 @@ Combining hashes/dictionaries
+    {{ a | combine(b, c, d) }}
+    {{ [a, b, c, d] | combine }}
+
+In this case, keys in `d` would override those in `c`, which would
```
```suggestion
In this case, keys in ``d`` would override those in ``c``, which would
```
This is Restructured Text, so use double backticks for code formatting.

tchernomax replied:
done (plus all the other ` in the combine section)

felixfontein commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -299,25 +299,36 @@ def mandatory(a, msg=None):
+    recursive = kwargs.pop('recursive', False)
+    list_merge = kwargs.pop('list_merge', 'replace')
+    if kwargs:
+        raise AnsibleFilterError("'recursive' and 'list_merge' are the only valid keyword argument")
```
```suggestion
        raise AnsibleFilterError("'recursive' and 'list_merge' are the only valid keyword arguments")
```

tchernomax replied:
done

tchernomax committed:
combine filter: spelling fix

tchernomax commented:
@felixfontein thanks for the review.
I have commited your fixes.

alexz-kh commented:
any updates around ?

tchernomax commented:
@alexz-kh I would really like this PR to be merged… but I can only regularly ask for review/merge on IRC and wait…
If you are interested by this PR please review it and if you think it's ready : write a "[shipit](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#commands)" comment

felixfontein commented on docs/docsite/rst/user_guide/playbooks_filters.rst:
```diff
@@ -890,36 +890,217 @@ Combining hashes/dictionaries
+
+list_merge
+  Is a string, it's possible values are ``replace`` (default), ``keep``, ``append``, ``prepend``, ``append_rp`` or ``prepend_rp``.
+  It modify the behaviour of ``combine`` when the hashes to merge contains arrays/lists.
```
```suggestion
  It modifies the behaviour of ``combine`` when the hashes to merge contain arrays/lists.
```

tchernomax replied:
done

felixfontein commented on docs/docsite/rst/user_guide/playbooks_filters.rst:
```diff
@@ -890,36 +890,217 @@ Combining hashes/dictionaries
+      - default
+
+If ``list_merge='append_rp'``, arrays from the right hash will be appended to the ones in the left hash.
+And elements of the left hash's array that are also in the right hash's array will be remove ("rp" stands for "remove present").
```
```suggestion
Elements of arrays in the left hash that are also in the corresponding array of the right hash will be removed ("rp" stands for "remove present").
```

tchernomax replied:
done

felixfontein commented on docs/docsite/rst/user_guide/playbooks_filters.rst:
```diff
@@ -890,36 +890,217 @@ Combining hashes/dictionaries
+      - 5
+      - 5
+
+If ``list_merge='prepend_rp'``, is the same as ``append_rp`` but elements of the right hash's array are prepend::
```
```suggestion
If ``list_merge='prepend_rp'``, the behavior is similar to the one for ``append_rp``, but elements of arrays in the right hash are prepended::
```

tchernomax replied:
done

felixfontein commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -45,7 +45,7 @@
 from ansible.module_utils._text import to_bytes, to_native, to_text
 from ansible.module_utils.common.collections import is_sequence
-from ansible.module_utils.common._collections_compat import Mapping, MutableMapping
+from ansible.module_utils.common._collections_compat import Mapping, MutableMapping, MutableSequence
```
I don't see where in the code you're using this.

tchernomax replied:
Indeed, must be a left over from my tests/development.
I also removed MutableMapping which wasn't used either.

felixfontein commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -299,25 +299,36 @@ def mandatory(a, msg=None):
+        raise AnsibleFilterError("'recursive' and 'list_merge' are the only valid keyword arguments")
+
+    # allow the user to do `[dict1, dict2, ...] | combine`
+    dictionaries = flatten(terms, levels=1)
```
Why not leave `levels=1` away? From your PR description, this breaks backwards compatibility for no reason (or if there's a reason, it's hidden somewhere).

tchernomax replied:
> non backward compatible: `[dict1, [dict2]] | combine` doesn't work anymore, because I thought it shouldn't. If you disagree, `flatten(terms, levels=1)` → `flatten(terms)` and it's fixed.

There is no technical reason.
Without this PR, `[dict1, [dict2]] | combine` is working but it's an undocumented behavior. As an ansible user, when I provide bad input to a filter I expect it to raise an error instead of guessing what I want to achieve.
Providing `[dict1, [dict2]]` to `combine` is a bad input as nothing in the documentation state it will flatten the list.

Though it's up for debate and if you disagree I will gradely remove the `levels=1`. I don't want that to block the entire PR.

@felixfontein What do you think ?

felixfontein replied:
If this is a bug that's fixed, there should be a changelog fragment (with `bugfixes:`) that mentions that this was a bug that is now fixed. I am fine if this is fixed / the undocumented behavior removed, but it should be clearly documented.

(If someone else complains, maybe move it into a separate PR, resp. a follow-up PR. If nobody else complains, just leave it there - assuming there is a bugfix changelog fragment for thix.)

tchernomax replied:
Ok, thanks !
I [created a changelog fragment](https://github.com/ansible/ansible/pull/57894/commits/3c323c1b0c1d803794620a57105d4b3482ba74b4):
```
+bugfixes:
+- combine filter - ``[dict1, [dict2]] | combine`` now raise an error; previously ``combine`` had an undocumented behaviour where it was flattening the list before combining it (https://github.com/ansible/ansible/pull/57894#discussion_r339517518).
+minor_changes:
+- combine filter - now accept a ``list_merge`` argument which modifies its behaviour when the hashes to merge contain arrays/lists.
```

felixfontein commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -299,25 +299,36 @@ def mandatory(a, msg=None):
+    if len(dictionaries) == 1:
+        return dictionaries[0]
+
+    # merge all the dicts so that the dict at the end of the array have precedence
```
If you write complete sentences, please use correct casing.
```suggestion
    # Merge all the dicts so that the dict at the end of the array have precedence
```

tchernomax replied:
done

felixfontein commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -299,25 +299,36 @@ def mandatory(a, msg=None):
+
+    # merge all the dicts so that the dict at the end of the array have precedence
+    # over the dict at the beginning.
+    # we merge the dicts from the highest to the lowest priority because there is
```
```suggestion
    # We merge the dicts from the highest to the lowest priority because there is
```

tchernomax replied:
done

felixfontein commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -299,25 +299,36 @@ def mandatory(a, msg=None):
+    # merge all the dicts so that the dict at the end of the array have precedence
+    # over the dict at the beginning.
+    # we merge the dicts from the highest to the lowest priority because there is
+    # a huge probability that the lowest priority dict will be the biggest in size
```
Do you have any proof that the probability is huge, or are you just claiming that? If you don't have a proof, how about writing "[...] because we're assuming that [...]"?

tchernomax replied:
> Do you have any proof that the probability is huge, or are you just claiming that?

It's just a claim assuming the lowest priority dict will be biggest, assuming it will contain default value, assuming there will be more default values than overwrites.
So indeed → your wording is better.

felixfontein commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -299,25 +299,36 @@ def mandatory(a, msg=None):
+    # a huge probability that the lowest priority dict will be the biggest in size
+    # (as the low prio dict will hold the "default" values and the others will be "patches")
+    # and merge_hash create a copy of it's first argument.
+    # so high/right -> low/left is more efficient than low/left -> high/right
```
```suggestion
    # So high/right -> low/left is probably more efficient than low/left -> high/right.
```

tchernomax replied:
done

felixfontein commented on lib/ansible/utils/vars.py:
```diff
@@ -92,32 +92,92 @@ def combine_vars(a, b):
-
-    return result
+    if list_merge not in ('replace', 'keep', 'append', 'prepend', 'append_rp', 'prepend_rp'):
+        raise AnsibleError("merge_hash: 'list_merge' argument can only be equal to 'replace', 'keep', 'append', 'prepend', 'append_rp' or 'prepend_rp'")
```
```suggestion
        raise AnsibleError("merge_hash: 'list_merge' argument must be one of 'replace', 'keep', 'append', 'prepend', 'append_rp' and 'prepend_rp'")
```

tchernomax replied:
done

felixfontein commented on lib/ansible/utils/vars.py:
```diff
@@ -92,32 +92,92 @@ def combine_vars(a, b):
+        x.update(y)
+        return x
+
+    # insert each element of y in x, overriding the one in x
```
If you have a comment spanning over multiple lines, consisting of several sentences, please use correct casing and terminate sentences with a period.

tchernomax replied:
done

felixfontein commented on lib/ansible/utils/vars.py:
```diff
@@ -92,32 +92,92 @@ def combine_vars(a, b):
+                # replace x value by y's one as it has higher priority
+                x[key] = y_value
+            elif list_merge == 'append':
+                x[key] += y_value
```
You need to make a copy of `x[key]` before you add some more elements:
```.py
>>> a = {1:[2]}
>>> b = dict(a)
>>> b[1] += [3]
>>> a
{1: [2, 3]}
```
Alternatively, write `x[key] = x_value + y_value`.

tchernomax replied:
Good catch !
fixed.

felixfontein commented on lib/ansible/utils/vars.py:
```diff
@@ -92,32 +92,92 @@ def combine_vars(a, b):
+    # insert each element of y in x, overriding the one in x
+    # (as y has higher priority)
+    # we copy elements from y to x instead of x to y because
+    # there is a high probability x will be the "default" dict the user
```
Same here: I don't think the probability can be proven to be high, but that you're making an assumption here. So please state that you're making an assumption.

tchernomax replied:
done

tchernomax committed:
filter/core: remove unused import

tchernomax committed:
merge_hash: fix x argument modification

```
>>> x=init={1:[2]}
>>> x=x.copy()
>>> x[1] += [3]
>>> init
{1: [2, 3]}
```

`+=` doesn't copy the list before modifying it.
So by user `+=` we where modifying lists inside first parameter of merge_hash.

tchernomax commented:
@felixfontein thanks for the detailed review ! I did all the modifications you ask.
Just one question left: https://github.com/ansible/ansible/pull/57894#discussion_r339517518

@bcoca I fixed what you ask earlier but github keep stating that "bcoca requested changes".
Can you re-review ? or validate my fix ? or something :)

Thanks

felixfontein commented on docs/docsite/rst/user_guide/playbooks_filters.rst:
```diff
@@ -890,36 +890,217 @@ Combining hashes/dictionaries
+  Note: It does **not** depend on the value of the ``hash_behaviour`` setting in ``ansible.cfg``.
+
+list_merge
+  Is a string, it's possible values are ``replace`` (default), ``keep``, ``append``, ``prepend``, ``append_rp`` or ``prepend_rp``.
```
```suggestion
  Is a string, its possible values are ``replace`` (default), ``keep``, ``append``, ``prepend``, ``append_rp`` or ``prepend_rp``.
```

tchernomax replied:
done

felixfontein commented:
This needs a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to).

tchernomax committed:
combine filter: changelogs fragment

tchernomax commented:
> This needs a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to).

done

tchernomax commented:
Can someone relaunch the CI, I think the fail has nothing to do with my commit.

tchernomax committed:
combine filter: documentation

tchernomax commented:
Can someone relaunch the CI, I think the fail has nothing to do with my commit.

felixfontein commented:
@tchernomax done

samccann commented:
@tchernomax can you please rebase so I can review the docs portion of this PR?  thanks!

tchernomax commented:
Can someone relaunch the CI, I think the fail has nothing to do with my commit.

ping @samccann

felixfontein commented:
CI restarted.

tchernomax commented:
thanks felix

@samccann can you review ?

(it would be awesome to see this one year old PR moving forward !)

felixfontein reviewed (approved):
LGTM

samccann reviewed (approved):
Docs LTGM

samccann commented:
Thanks @tchernomax for your patience on this Ansible fix!
